File: getHistory.go
package main

import (
	"encoding/csv"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"sort"
	"strconv"
	"time"

	"github.com/sirupsen/logrus"
)

type Option struct {
	Symbol    string `json:"symbol"`
	Interval  string `json:"interval"`
	StartTime int64  `json:"startTime"`
	EndTime   int64  `json:"endTime"`
}

func main() {
	now := time.Now().Unix() * 1000
	duration := 3600 * 1000 // 3600 seconds
	//temp
	now = 1715747119000
	startTime := now - int64(duration)
	endtime := now
	skipHeader := false
	for {

		// init data
		options := Option{
			Symbol:    "btcthb",
			Interval:  "1m",
			StartTime: startTime,
			EndTime:   endtime,
		}
		params := map[string]interface{}{}
		dataByte, _ := json.Marshal(options)
		json.Unmarshal(dataByte, &params)

		qs := constructQueryStringWithPrefix(params, "")
		url := "https://api.binance.th/api/v1/klines"

		millis := int64(startTime)
		// Convert milliseconds to seconds and nanoseconds
		seconds := millis / 1000
		nanoseconds := (millis % 1000) * int64(time.Millisecond)
		timestamp := time.Unix(seconds, nanoseconds)
		formattedTime := timestamp.Format("02/01/2006 15:04:05")
		fmt.Println("\nstart time", formattedTime, " call ", url+"?"+qs)
		req, err := http.NewRequest("GET", url+"?"+qs, nil)
		if err != nil {
			fmt.Println(err.Error())

		}

		req.Header.Add("Content-Type", "application/json")
		clientHttp := http.Client{}
		resp, _ := clientHttp.Do(req)
		body, err := ioutil.ReadAll(resp.Body)
		if err != nil {
			fmt.Println(err.Error())

		}
		if resp.StatusCode != 200 {
			fmt.Println(resp)
			break
		}
		fmt.Println(resp.Status)

		//write file
		var dataObj [][]interface{}
		json.Unmarshal(body, &dataObj)
		writeFile(dataObj, skipHeader)
		skipHeader = true
		endtime = startTime
		startTime = endtime - int64(duration)

		time.Sleep(time.Second * 2)
	}

}

func constructQueryStringWithPrefix(params map[string]interface{}, prefix string) string {
	var keys []string

	for k, _ := range params {
		keys = append(keys, k)
	}

	sort.Strings(keys)

	var qs string

	for i, k := range keys {
		v := params[k]

		if nestedParams, ok := v.(map[string]interface{}); ok {
			qs += constructQueryStringWithPrefix(nestedParams, k)
		} else if array, ok := v.([]interface{}); ok {
			logrus.Errorf("bar")
			nestedMap := map[string]interface{}{}
			for i, v := range array {
				nestedMap[fmt.Sprintf("%d", i)] = v
			}
			qs += constructQueryStringWithPrefix(nestedMap, k)
		} else {
			if prefix == "" {
				if _, ok := v.(float64); ok {
					qs += fmt.Sprintf("%s=%.f", k, v)
				} else {
					qs += fmt.Sprintf("%s=%v", k, v)
				}
			} else {
				if _, ok := v.(float64); ok {
					qs += fmt.Sprintf("%s[%s]=%.f", prefix, k, v)
				} else {
					qs += fmt.Sprintf("%s[%s]=%v", prefix, k, v)
				}
			}
		}

		if i != len(keys)-1 {
			qs += "&"
		}
	}

	return qs
}

func writeFile(dataObjs [][]interface{}, skipHeader bool) {

	file, err := os.OpenFile("klines.csv", os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		fmt.Println("Error opening the file:", err)
		return
	}
	defer file.Close()

	// Create a CSV writer
	writer := csv.NewWriter(file)
	defer writer.Flush()
	// Write the CSV header
	if !skipHeader {
		header := []string{"Open time", "Open", "High", "Low", "Close",
			"Volume", "Close time", "Quote asset volume", "Number of trades", "Taker buy base asset volume", "Taker buy quote asset volume"}

		err = writer.Write(header)
		if err != nil {
			panic(err)
		}
	}
	// Write each struct instance as a CSV row
	for _, data := range dataObjs {

		openTime := strconv.FormatFloat(data[0].(float64), 'f', 0, 64)
		closeTime := strconv.FormatFloat(data[6].(float64), 'f', 0, 64)
		NoT := strconv.FormatFloat(data[8].(float64), 'f', 0, 64)

		row := []string{openTime, data[1].(string), data[2].(string), data[3].(string), data[4].(string),
			data[5].(string), closeTime, data[7].(string), NoT, data[9].(string), data[10].(string),
		}
		err := writer.Write(row)
		if err != nil {
			panic(err)
		}
	}
}

File: get_history.py
import csv
import json
import time
import requests
from datetime import datetime

class Option:
    def __init__(self, symbol, interval, start_time, end_time):
        self.symbol = symbol
        self.interval = interval
        self.start_time = start_time
        self.end_time = end_time

def construct_query_string(params, prefix=''):
    keys = sorted(params.keys())
    qs = ''
    for i, k in enumerate(keys):
        v = params[k]
        if isinstance(v, dict):
            qs += construct_query_string(v, k)
        elif isinstance(v, list):
            nested_map = {str(i): v[i] for i in range(len(v))}
            qs += construct_query_string(nested_map, k)
        else:
            if prefix:
                qs += f"{prefix}[{k}]={v}"
            else:
                qs += f"{k}={v}"
        if i != len(keys) - 1:
            qs += "&"
    return qs

def write_file(data_objs, skip_header):
    with open('klines.csv', mode='a', newline='') as file:
        writer = csv.writer(file)
        if not skip_header:
            header = ["Open time", "Open", "High", "Low", "Close", "Volume", 
                      "Close time", "Quote asset volume", "Number of trades", 
                      "Taker buy base asset volume", "Taker buy quote asset volume"]
            writer.writerow(header)
        for data in data_objs:
            row = [
                str(data[0]), data[1], data[2], data[3], data[4],
                data[5], str(data[6]), data[7], str(data[8]), data[9], data[10]
            ]
            writer.writerow(row)

def main():
    now = int(time.time() * 1000)
    duration = 3600 * 1000
    now = 1715747119000
    start_time = now - duration
    end_time = now
    skip_header = False
    
    while True:
        options = Option("btcthb", "1m", start_time, end_time)
        params = options.__dict__
        qs = construct_query_string(params)
        url = "https://api.binance.th/api/v1/klines"
        
        millis = start_time
        seconds = millis // 1000
        nanoseconds = (millis % 1000) * 1000000
        timestamp = datetime.fromtimestamp(seconds + nanoseconds / 1e9)
        formatted_time = timestamp.strftime("%d/%m/%Y %H:%M:%S")
        print(f"\nstart time {formatted_time} call {url}?{qs}")

        response = requests.get(f"{url}?{qs}")
        if response.status_code != 200:
            print(response)
            break
        print(response.status_code)
        
        data_obj = json.loads(response.content)
        write_file(data_obj, skip_header)
        skip_header = True
        end_time = start_time
        start_time = end_time - duration

        time.sleep(2)

if __name__ == "__main__":
    main()
